src/hotspot/share/runtime/sharedRuntime.cpp

```diff
@@ -677,6 +677,7 @@ JRT_END
 
 // ret_pc points into caller; we are returning caller's exception handler
 // for given exception
+// Note that the implementation of this method assumes it's only called when an exception has actually occured
 address SharedRuntime::compute_compiled_exc_handler(CompiledMethod* cm, address ret_pc, Handle& exception,
```

Comment:
One thing of note for this function: We don't look up the bci for JVMCI compiled methods, so I've not added any profiling in the case of JVMCI (see the `#if INCLUDE_JVMCI` block at the start of the function).

This means that when using a JVMCI compiler, exception handlers might appear as untaken, when they are actually taken. This should be fine since the profiling information is currently only used by C2. But, if a JVMCI compiler wants to start using the profiling information e.g. to prune dead exception handlers as well, then profiling needs to be implement first.

Please let me know if this is okay.

Reply:
I think that's fine (@dougxc, fyi).

Reply:
I also think it's fine but would like @tkrodriguez to confirm as well.

Reply:
I think it should be straightforward to add this profile for JVMCI compiler at line 693 with
```
      if (t->bci() != -1) { // did we find a handler in this method?
        sd->method()->set_ex_handler_entered(t->bci()); // profile
      }
```

Reply:
Note that we don't have an `sd` at that point in the code. `cm->method()` only points at the top-level method, so we'd have to do the same method/handler bci lookup as we do for C2 in the big `while` loop below the JVMCI code. Given that we don't do that already, I figured JVMCI exception handler routing just worked differently, and there was another point (on the Java side) where the method + bci is looked up.

Reply:
I think this fine for now as well.  The code as written just can't work for JVMCI since we don't even enter that part of the logic.  We do a straight lookup based on the pc offset and dispatch there directly.  The generated code actually does the exception dispatch for the inlined scopes so it's not visible to the runtime which handler actually services the request.